pkg/runner/runner.go

```diff
@@ -0,0 +1,205 @@
+/*
+Copyright 2019 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+*/
+
+package runner
+
+import (
+	"context"
+	"time"
+
+	"github.com/gravitational/fakeiot/pkg/client"
+	"github.com/gravitational/fakeiot/pkg/metric"
+	"github.com/pborman/uuid"
+
+	"github.com/gravitational/trace"
+	log "github.com/sirupsen/logrus"
+)
+
+// Config is a runner configuration
+type Config struct {
+	// Client is a configured client
+	Client *client.Client
+}
+
+// New returns a new instance of runner
+func New(config Config) *Runner {
+	return &Runner{
+		Config: config,
+		Entry:  log.WithFields(log.Fields{trace.Component: "runner"}),
+	}
+}
+
+// Runner is Fake IOT device runner
+type Runner struct {
+	// Config is a device configuration
+	Config
+	*log.Entry
+}
+
+// Simulation is a IOT device simulation
+type Simulation struct {
+	// Period is a period to run simulation for
+	Period time.Duration
+	// Freq is a frequency of activity
+	Freq time.Duration
+	// AccountID is an account ID to emit
+	AccountID string
+	// Users is a total amount of different users to simulate accross
+	// all accounts. Simulation will generate distinct user IDs
+	// and emit them in a rotating loop
+	Users int
+}
+
+// RunSimulation runs test simulation
+func (r *Runner) RunSimulation(ctx context.Context, sim Simulation) error {
+	r.Infof("Starting simulation with %v users over %v period with %v frequency.", sim.Users, sim.Period, sim.Freq)
+	tctx, cancel := context.WithTimeout(ctx, sim.Period+time.Second)
+	defer cancel()
+
+	var users []string
+	for i := 0; i < sim.Users; i++ {
+		users = append(users, uuid.New())
+	}
+
+	t := time.NewTicker(sim.Freq)
+	defer t.Stop()
+	var i int
+	for {
+		user := users[i]
+		i = (i + 1) % len(users)
+		metric := metric.Metric{
+			AccountID: sim.AccountID,
+			UserID:    user,
+			Timestamp: time.Now().UTC(),
+		}
+		err := r.Client.Send(tctx, metric)
+		if err != nil {
+			return trace.Wrap(err)
```

Comment:
I'm just curious, does it fail on the first error on purpose or would it be better to let it run for the specified simulation duration and collect all errors? I guess it depends on what the purpose of this simulation is.

Reply:
yes, it fails on purpose